R/mlr_graphs_elements.R

```diff
@@ -0,0 +1,163 @@
+
+  cols_by_type = function(types) {
+    task$feature_types[get("type") %in% types,]
+  }
+  if (!is.null(task)) assert_task(task)
+  if (!is.null(learner)) assert_learner(learner)
+  if (is.null(impute_missings)) impute_missings = (is.null(task) || any(task$missings())) && (is.null(learner) || !"missings" %in% learner$properties)
+  else assert_flag(impute_missings)
+  if (is.null(factors_to_numeric)) factors_to_numeric = (is.null(task) || nrow(cols_by_type("factor")) > 0) && (is.null(learner) || !"factors" %in% learner$properties)
+  else assert_flag(impute_missings)
+  assert_flag(scaling)
+  assert_count(max_cardinality)
+
+  # FIXME
+  # This should additionally do the following thins:
+  # - Convert posixct to numeric using new pipeop
+  # - Auto-convert character to factor? Perhaps improve later with text po's
+  # - Logical to integer conversion
+
+  # If given a task, only treat actually existing column types
+  if (!is.null(task)) {
+    pos = list()
+    if ("factor" %in% learner$feature_types && nrow(cols_by_type("factor")) > 0)
+      pos = c(pos, po("fixfactors", param_vals = list(affect_columns = is.factor)))
+    if (impute_missings) {
+      # Impute numerics
+      if (nrow(cols_by_type(c("numeric", "integer"))) > 0)
+        pos = c(pos, po("copy", 2) %>>%
+          gunion(list(
+            po("imputehist", param_vals = list(affect_columns = selector_type(c("numeric", "integer")))),
+            po("missind",  param_vals = list(affect_columns = selector_type(c("numeric", "integer")))))) %>>%
+          po("featureunion"))
+      # Impute factors
+      if (nrow(cols_by_type(c("factor", "ordered"))) > 0)
+        pos = c(pos, po("imputenewlvl", param_vals = list(affect_columns = selector_type(c("factor", "ordered")))))
+    }
+
+    if (factors_to_numeric) {
+      # Collapse factors over 1000 levels
+      if (any(map_lgl(task$levels(cols_by_type("factor")$id), function(x) length(x) > max_cardinality)))
+        pos = c(pos, po("collapsefactors", param_vals = list(target_level_count = max_cardinality)))
+      # Encode factors
+      pos = c(pos, po("encode"))
+    }
+  # If no task is provided, be conservative
+  } else {
+    pos = list(po("fixfactors", param_vals = list(affect_columns = selector_type(c("factor", "ordered")))))
+    if (impute_missings) {
+      pos = c(pos,
+        po("copy", 2) %>>%
+          gunion(list(
+            po("imputehist", param_vals = list(affect_columns = selector_type(c("numeric", "integer")))),
+            po("missind",  param_vals = list(affect_columns = selector_type(c("numeric", "integer")))))) %>>%
+          po("featureunion"),
+        po("imputenewlvl", param_vals = list(affect_columns = selector_type(c("factor", "ordered")))))
+    }
+    pos = c(pos, po("collapsefactors", param_vals = list(target_level_count = max_cardinality)))
+    if (factors_to_numeric) pos = c(pos, po("encode"))
+  }
+
+  if (scaling) pos = c(pos, po("scale"))
```

Comment:
is scaling every necessary to make a model robust against crashes? Otherwise this becomes a generic preprocessing pipeline.

Reply:
I think we might wanna use robustify as something I can plug in front of an `svm` and get sensible results. 
I think I will just have `scaling` default to `FALSE` though.